Update external_position_estimation.md (#876)
PX4/Devguide: en/ros/external_position_estimation.md

## Reference Frames

PX4 uses FRD (X **F**orward, Y **R**ight and Z **D**own) for the local body frame as well for the reference frame. When using the heading of the magnetometer, the PX4 reference frame x axis will be aligned with north, so therefore it is called NED (X **N**orth, Y **E**ast, Z **D**own). The heading of the reference frame of the PX4 estimator and the one of the external pose estimate will not match in most cases. Therefore the reference frame of the external pose estimate is named differently, it is called [MAV_FRAME_LOCAL_FRD](https://mavlink.io/en/messages/common.html#MAV_FRAME_LOCAL_FRD).

> Depending on the source of your reference frame, you will need to apply a custom transformation to the pose estimate before sending the MAVLink Vision/MoCap message. This is necessary to change the orientation of the parent and child frame of the pose estimate, such that it fits the PX4 convention. Have a look at the MAVROS [odometry plugin](https://github.com/mavlink/mavros/blob/master/mavros_extras/src/plugins/odom.cpp) for the necessary transformations.

Reviewer: ```suggestion
Depending on the source of your reference frame, you will need to apply a custom transformation to the pose estimate before sending the MAVLink Vision/MoCap message. This is necessary to change the orientation of the parent and child frame of the pose estimate, such that it fits the PX4 convention. Have a look at the MAVROS [*odom* plugin](https://github.com/mavlink/mavros/blob/master/mavros_extras/src/plugins/odom.cpp) for the necessary transformations.
```

Reviewer: Some images here exposing the transforms would look quite more interesting than just pointing to code in MAVROS.

Author: I agree, but I can't spend time on it.

Both frames are shown in the image below (FLU on left/FRD on right).

![Reference frames](../../assets/lpe/ref_frames.png)

With EKF2 when using external heading estimation, magnetic north can either be ignored and or the heading offset to magnetic north can be calculated and compensated. Depending on your choice the yaw angle is given with respect to either magnetic north or local *x*.

> **Note** When creating the rigid body in the MoCap software, remember to first align the robot's local *x* axis with the world *x* axis otherwise the yaw estimate will have an offset. This can stop the external pose estimate fusion from working properly. Yaw angle should be zero when body and reference frame align.

Using MAVROS, this operation is straightforward. 
ROS uses ENU frames as convention, therefore position feedback must be provided in ENU. 
If you have an Optitrack system you can use [mocap_optitrack](https://github.com/ros-drivers/mocap_optitrack) node which streams the object pose on a ROS topic already in ENU. 
With a remapping you can directly publish it on `mocap_pose_estimate` as it is without any transformation and MAVROS will take care of NED conversions.

> The MAVROS odometry plugin makes it easy to handle the coordinate frames. It uses ROS's tf package. Your external pose system might have a completely different frame convention that does not match the one of PX4. The body frame of the external pose estimate can depend on how you set the body frame in the MOCAP software or on how you mount the VIO sensor on the drone. The MAVROS odometry plugin needs to know how the external pose's child frame is oriented with respect to either the FCU's FRD or FLU body frame known by MAVROS. You therefore have to add the external pose's body frame to the tf tree. This can be done by including an adapted version of the following line into your ROS launchfile. 

Reviewer: Let's change the naming *FCU*  to *aircraft*.

Reviewer: And should be ROS launch file, instead of *launchfile*.

```
  <node pkg="tf" type="static_transform_publisher" name="tf_baseLink_externalPoseChildFrame"
        args="0 0 0 <yaw> <pitch> <roll> base_link <external_pose_child_frame> 1000"/>
```

> Make sure that you change the values of yaw, pitch and roll such that it properly attaches the external pose's body frame to the `base_link` or `base_link_frd`. Have a look at the [tf package](http://wiki.ros.org/tf#static_transform_publisher) for further help on how to specify the transformation between the frames. You can use rviz to check if you attached the frame right. The name of the `external_pose_child_frame` has to match the child_frame_id of your `nav_msgs/Odometry` message.

Reviewer: ```suggestion
Make sure that you change the values of the orientation such that it properly attaches the external pose's body frame to the `base_link`. Have a look at the [tf package](http://wiki.ros.org/tf#static_transform_publisher) for further help on how to specify the transformation between the frames. You can use *Rviz* to check if you attached the frame right. The name of the `external_pose_child_frame` has to match the child_frame_id of your [`nav_msgs/Odometry`](http://docs.ros.org/melodic/api/nav_msgs/html/msg/Odometry.html) message.
```

> The same also applies for the reference frame of the external pose. You have to attach the reference frame of the external pose as child to the either the `odom` or `odom_frd` frame. Adapt therefore the following code line accordingly.

Reviewer: ```suggestion
The same also applies for the reference frame of the external pose. You have to attach the reference frame of the external pose as child to the `odom` frame. Adapt therefore the following code line accordingly.
```

Author: You can connect it to either frame

```
  <node pkg="tf" type="static_transform_publisher" name="tf_odom_externalPoseParentFrame"
        args="0 0 0 <yaw> <pitch> <roll> odom <external_pose_parent_frame> 1000"/>
```
If the reference frame has the z axis pointing upwards you can attached it without any rotation (yaw=0, pitch=0, roll=0) to the `odom` frame. The name of `external_pose_parent_frame` has to match the frame_id of the odometry message.

> > **Note** When using the MAVROS odometry plugin, it is important that nobody is publishing a tf transformation between the external pose's reference and child frame. This might break the tf tree.

Reviewer: ```suggestion
> **Note** When using the MAVROS *odom* plugin, it is important that no other node is publishing a transform between the external pose's reference and child frame. This might break the *tf* tree.
```

After setting up one of the (specific) systems described above you should now be ready to test.
The instructions below show how to do so for MoCap and VIO systems 

### Check external estimate

Be sure to perform the following checks before your first flight:

> * Set the PX4 parameter `MAV_ODOM_LP` to 1. PX4 will therefore publish the received external pose on the MAVLink [ODOMETRY](https://mavlink.io/en/messages/common.html#ODOMETRY) topic.

Reviewer: ```suggestion
* Set the PX4 parameter `MAV_ODOM_LP` to 1. PX4 will therefore stream back the received external pose as MAVLink [ODOMETRY](https://mavlink.io/en/messages/common.html#ODOMETRY) messages.
```

> * Check this MAVLink message with e.g. the Analyze Widget of *QGroundControl*. In order to do this, yaw the vehicle until the quaternion of the ODOMETRY message is very close to a unit quaterion. (w=1, x=y=z=0)

Reviewer: ```suggestion
* It is recommended to check these MAVLink messages with e.g. the *Analyze* Widget of *QGroundControl*. In order to do this, yaw the vehicle until the quaternion of the ODOMETRY message is very close to a unit quaternion. (w=1, x=y=z=0)
```

* At this point the body frame is aligned with the reference frame of the external pose system. If you do not manage to get a quaternion close to the unit quaternion without rolling or pitching your vehicle, your frame probably still have a pitch or roll offset. Do not proceed if this is the case and check your coordinate frames again.

> * Once aligned you can pick the vehicle up from the ground and you should see the position's z coordinate decrease. Moveing the vehilce in forward direction, should increase the position's x coordinate. While moving the vehicle to the right should increase the y coordinate. In the case you send also linear velocities from the external pose system, you should also check the linear velocities. Check that the linear velocities are in expressed in the FRD body frame.

Reviewer: ```suggestion
* Once aligned you can pick the vehicle up from the ground and you should see the position's z coordinate decrease. Moving the vehicle in forward direction, should increase the position's x coordinate. While moving the vehicle to the right should increase the y coordinate. In the case you send also linear velocities from the external pose system, you should also check the linear velocities. Check that the linear velocities are in expressed in the *FRD* body frame reference frame.
```

> * Set the PX4 parameter `MAV_ODOM_LP` back to 0. Therefore the ODOMETRY message will contain now the fused estimate coming from the estimator.

Reviewer: ```suggestion
* Set the PX4 parameter `MAV_ODOM_LP` back to 0. PX4 will stop streaming this message back.
```

Increase the value of the left stick and the robot will take off, 
put it back to the middle right after. Check if it is able to keep its position.

> <<<<<<< HEAD

Reviewer 2: Cruft here!

Author: Sry, I missed that.

put it back to the middle right after. Check if it is able to keep its position.

If it works, you may want to set up an [offboard](offboard_control.md) experiment by sending position-setpoint from a remote ground station.

> ### VIO First Flight

Reviewer: This needs to eventually get in. Do not remove it so it stays here as a reference of something that needs to be done.

Author: Added it back
